File: src/decoders/meal.ts
import type { Meal, MealCategory } from "~/models";

import { decodeMealCategory } from "./meal-category";
import { decodeMoment } from "./moment";

export function decodeMeal (meal: any): Meal {
  let information: string | undefined;
  const categories = meal.foodcategory.map(decodeMealCategory)
    .filter((category: MealCategory) => {
      if (category.name === "informations" || category.name === "Fermeture") {
        if (typeof information === "undefined") {
          information = category.dishes[0];
        }

        return false;
      }

      return true;
    });

  return {
    moment: decodeMoment(meal.name),
    information,
    categories
  };
};

File: library/src/commonMain/kotlin/decoders/meal.kt
package decoders

import kotlinx.serialization.json.*
import models.Meal

fun decodeMeal (meal: JsonObject): Meal {
  var information: String? = null
  val categories = meal["foodcategory"]!!.jsonArray.map {
    decodeMealCategory(it.jsonObject)
  } .filter {
    if (it.name == "informations" || it.name == "Fermeture") {
      if (information == null) {
        information = it.dishes.first()
      }

      // skip this category
      false
    }

    else true
  }


  return Meal(
    moment = decodeMoment(meal["name"]!!.jsonPrimitive.content),
    information,
    categories
  )
}